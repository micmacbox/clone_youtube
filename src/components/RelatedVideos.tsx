import { useYoutubeApi } from "../context";
import { useQuery } from "@tanstack/react-query";
import { FC } from "react";
import VideoCard from "./VideoCard";
import { VideoItem } from "../@types/video";
import { VideoType } from "../constants";

type Props = {
  id: string;
};

const RelatedVideos: FC<Props> = ({ id }) => {
  const youtube = useYoutubeApi();
  const {
    isLoading,
    error,
    data: videos,
  } = useQuery({
    queryKey: ["channel", id],
    queryFn: () => youtube.channelVideos(id),
    staleTime: 1000 * 60 * 5,
  });
  return (
    <>
      {isLoading && <p>loading...</p>}
      {error && <p>error occured ðŸ˜°</p>}
      {videos && (
        <ul>
          {videos.map((video: VideoItem) => (
            <VideoCard
              key={video.id}
              video={video}
              type={VideoType.relatedVideos}
            />
          ))}
        </ul>
      )}
    </>
  );
};

export default RelatedVideos;
