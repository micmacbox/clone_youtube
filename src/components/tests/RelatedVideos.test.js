import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { Route } from "react-router-dom";
import { fakeVideos } from "../../test";
import { withAllContexts, withRouter } from "../../test/utils";
import RelatedVideos from "../RelatedVideos";

describe("RelatedVideos", () => {
  const fakeYoutube = {
    channelVideos: jest.fn(),
  };

  afterEach(fakeYoutube.channelVideos.mockReset());

  it("renders correctly", async () => {
    fakeYoutube.channelVideos.mockImplementation(() => fakeVideos);
    const { asFragment } = renderRelatedVideos();

    await waitForElementToBeRemoved(screen.getByText("loading..."));
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders related videos correctly", async () => {
    fakeYoutube.channelVideos.mockImplementation(() => fakeVideos);
    renderRelatedVideos();

    expect(fakeYoutube.channelVideos).toHaveBeenCalledWith("id");
    await waitFor(() =>
      expect(screen.getAllByRole("listitem")).toHaveLength(fakeVideos.length)
    );
  });

  it("renders error", async () => {
    fakeYoutube.channelVideos.mockImplementation(() => {
      throw new Error("error");
    });

    renderRelatedVideos();
    await waitFor(() => {
      expect(screen.getByText("error occured ðŸ˜°")).toBeInTheDocument();
    });
  });

  it("renders loading", () => {
    fakeYoutube.channelVideos.mockImplementation(() => fakeVideos);
    renderRelatedVideos();

    expect(screen.getByText("loading...")).toBeInTheDocument();
  });

  function renderRelatedVideos() {
    return render(
      withAllContexts(
        withRouter(<Route path="/" element={<RelatedVideos id="id" />} />),
        fakeYoutube
      )
    );
  }
});
