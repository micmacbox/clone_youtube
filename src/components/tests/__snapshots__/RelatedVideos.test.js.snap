// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideos renders correctly 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="England vs. Italy Highlights | European Qualifiers"
        class="w-60 mr-2"
        src="https://i.ytimg.com/vi/LcKoyrvqE4I/mqdefault.jpg"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          England vs. Italy Highlights | European Qualifiers
        </p>
        <p
          class="text-sm opacity-80"
        >
           TUDN USA
        </p>
        <p
          class="text-sm opacity-80"
        >
          4 days ago
        </p>
      </div>
    </li>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="England vs. Italy Highlights | European Qualifiers2"
        class="w-60 mr-2"
        src="https://i.ytimg.com/vi/LcKoyrvqE4I/mqdefault2.jpg"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          England vs. Italy Highlights | European Qualifiers2
        </p>
        <p
          class="text-sm opacity-80"
        >
           TUDN USA
        </p>
        <p
          class="text-sm opacity-80"
        >
          4 days ago
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
